### ISSUE TITLE
Deleted branches are still showing up under branches as remote

### ISSUE THREAD

**Issue by famosab** [gitbutlerapp/gitbutler#4161]
### Version

0.12.5

### Operating System

Mac OS X

### Distribution Method

dmg (Apple Silicon)

### Describe the issue

![image](https://github.com/gitbutlerapp/gitbutler/assets/45968370/fab532c4-37c1-4309-b452-b4d0cc9c0cf9)
This is a branch that was merged and deleted upstream and locally. 

### Expected behavior

Deleted branches should not show up anymore.

**Comment by Byron** [comment]
Thanks for reporting!

Does this issue persist, that is, when reopening the project then deleted branches are still showing up?

I wonder if the "How to reproduce" involves deleting the branches while keeping the UI open, but the UI does not update accordingly?
Thanks again for clarifying.

**Comment by famosab** [comment]
Yes, I removed the project and added it again and all the branches are still showing up. 

My repository is a fork of an upstream to which I contribute regulary with Pull Requests.

**Comment by Byron** [comment]
Thanks a lot for elaborating. Now that it's clear that the issue persist, and knowing that GitButler can't make branches up, they must be there somewhere.

Can you find the branches it shows (directly or as suffixes, maybe parts of remote tracking branches) in the list you get with `git branch -a`? My thinking is that the branches are there under a different remote maybe, but shown in a truncated fashion in the UI, somehow.

**Comment by famosab** [comment]
These are the branches I get with `git branch -a`:
```
  cf-bed-circos
  cf-circos
  fix-strelka-somatic
* gitbutler/integration
  gitbutler/target
  manta-germ-test
  manta-tumor-test
  master
  strelka-germ-test
  strelka-somatic-test
  tiddit-sv
  update_mageck_test
  vcftools
  vep-dl
  remotes/origin/HEAD -> origin/master
  remotes/origin/getref-nftest
  remotes/origin/master
  remotes/origin/module-datavzrd
  remotes/origin/module-info
  remotes/origin/muse-call
  remotes/origin/muse-sump
```
And these are also the ones showing up in GitButler. They seem to be local branches (even though they are shown as remote) as they are not found on the remote repository. And when I try to delete them in GitButler they stay.

**Comment by Byron** [comment]
Going back to the original topic in the issue, `refs/heads/cf-circos` is supposed to be deleted locally, but it still shows in the listing above (as `cf-circos`), which means it exists. GitButler shows the full name. The remote tracking branch for it does not exist though, so maybe this is the source of the confusion?

Is there any other branch that is listed in the UI but not listed by `git branch -a`?

**Comment by famosab** [comment]
<img width="197" alt="image" src="https://github.com/gitbutlerapp/gitbutler/assets/45968370/735e12ac-8fa8-46d4-b7cd-99e0237a9198">
<img width="197" alt="image" src="https://github.com/gitbutlerapp/gitbutler/assets/45968370/c7f66ecc-68f7-4cf5-8891-86df6ac483e1">

The UI shows these 15 branches. I remember deleting for example cf-bed-circos, cf-circos and fix-strelka-somatic since the PRs were merged. So I deleted them on the repo and locally in the UI. After the update to 0.12.5 they showed up again.

**Comment by Byron** [comment]
Thank you!

Since GitButler sees these branches and Git sees these branches, to me it seems clear that they do exist.

But that's not what you are saying - what you *are* saying is that they have been re-created somehow when switching to a more recent GitButler version?

**Comment by famosab** [comment]
I deleted them in an older GitButler version and also on the remote repo. When I updated, these branches were showing up again in the UI. I never checked (before updating) whether they were present with `git branch -a` after deleting them in the UI (so maybe they were never really gone). Does that make senese?

**Comment by Byron** [comment]
I see - so the issue might be that the branches were never deleted by the UI in the first place. Maybe the steps were:

1. delete branches in GB pre v12.5
2. update GB
3. the new version v12.5 opens and shows the previously deleted branches

Since GitButler can't make these up (for instance by having its own configuration files with branch names), I think the branches weren't actually deleted on disk, even though the UI thought so.

Can imagine this really happened? Not really, the UI should have shown the error, but it's the best explanation of what might have happened. Maybe that's a way to reproduce the issue.

**Comment by famosab** [comment]
I am also not able to delete those branches via the UI anymore. I will go ahead and remove those using the command line.

**Comment by Byron** [comment]
Could it be that there is some sort of name-duplication with virtual branches? That way, you might have deleted virtual branches which somehow shadowed the real ones. I was also under the impression that real branches won't be touched by GitButler.

**Comment by famosab** [comment]
Sure all my virtual branches turn into real ones at some point when I commit and push.

**Comment by Byron** [comment]
Thanks for sharing!

I know that GItButler now creates real branches for everything it finds `.git/gitbutler/virtual_branches.toml`, so if the zombie-branch is mentioned there, there is reason to believe that GitButler could end up recreating it.

Could you check if this is the case?
Thanks again.

**Comment by LyricSeraph** [comment]
Having the same issue in version 0.20.4 GUI client on ArchLinux (using the prebuilt release on [AUR/gitbutler-bin](https://aur.archlinux.org/packages/gitbutler-bin) ).

It seems that the "fetch upstream" button on GUI only triggers a `git fetch`, instead of `git fetch --prune`, which would delete non-exist remote refs.

After running `git fetch --prune` in the terminal, the GUI shows a missing ref error on that branch. Then the branch will be disappeared after hitting "fetch upstream" again.

**Comment by Byron** [comment]
> It seems that the "fetch upstream" button on GUI only triggers a git fetch, instead of git fetch --prune, which would delete non-exist remote refs.

Chiming in only on one aspect: `--prune` was removed as it had other problems, I think it was less reliable. GB did use it before that. So it feels like maybe one day that should be an option/configurable.

> After running git fetch --prune in the terminal, the GUI shows a missing ref error on that branch. Then the branch will be disappeared after hitting "fetch upstream" again.

That's a bit strange, as git will only delete remote tracking refs with `--prune`, and these aren't usually displayed in the GUI. If they are, that's only possible in the branch list, and the underlying issue might be that the GUI doesn't yet update when references change underneath.
